sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionLifeCycleManager.cs

```diff
@@ -0,0 +1,248 @@
+                if (_threadCount == 0 || forceClose)
+                {
+                    await CloseSession(cancellationToken).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                if (releaseSemaphore)
+                {
+                    _semaphore.Release();
+                }
+            }
+        }
+
+        private async Task CloseSession(CancellationToken cancellationToken)
+        {
+            if (_receiver == null || _receiver.IsDisposed)
+            {
+                return;
+            }
+            try
+            {
+                if (_sessionCloseHandler != null)
+                {
+                    var args = new ProcessSessionEventArgs(_receiver, cancellationToken);
+                    await _sessionCloseHandler(args).ConfigureAwait(false);
+                }
+            }
+
+            catch (Exception exception)
+            {
+                await ProcessorUtils.RaiseExceptionReceived(
+                    _errorHandler,
+                    new ProcessErrorEventArgs(
+                        exception,
+                        ServiceBusErrorSource.CloseMessageSession,
+                        _fullyQualifiedNamespace,
+                        _entityPath))
+                    .ConfigureAwait(false);
+            }
+
+            finally
+            {
+                // Always at least attempt to dispose. If this fails, it won't be retried.
+                await _receiver.DisposeAsync().ConfigureAwait(false);
+                _receiver = null;
+            }
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="receiver"></param>
+        /// <param name="cancellationTokenSource"></param>
+        /// <returns></returns>
+        private async Task RenewSessionLock(
+            ServiceBusSessionReceiver receiver,
+            CancellationTokenSource cancellationTokenSource)
+        {
+            cancellationTokenSource.CancelAfter(_maxAutoLockRenewal);
+            CancellationToken cancellationToken = cancellationTokenSource.Token;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    ServiceBusEventSource.Log.ProcessorRenewSessionLockStart(_identifier, receiver.SessionId);
+                    TimeSpan delay = ProcessorUtils.CalculateRenewDelay(receiver.SessionLockedUntil);
+
+                    try
+                    {
+                        await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        break;
+                    }
+                    if (receiver.IsDisposed)
+                    {
+                        break;
+                    }
+                    await receiver.RenewSessionLockAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It is possible to get SessionLockLost exception when we renew session lock. Don't we want to call the user's error handler if we get SessionLockLost?

Reply:
Yeah we are calling it in the catch block as long as it is not ObjectDisposed. I just pushed a fix to prevent calling when the cancellation token is canceled.

Reply:
Ah.. I see. In other places you're using `ThrowIfSessionLockLost` method. Can we move `ThrowIfSessionLockLost` method in `ProcessorUtils` and use it here as well?

Reply:
We don't actually await the renewSessionLock task, so the exception wouldn't go anywhere. If the session lock is lost any individual operations on the receiver will throw.